flink/src/main/scala/ru/itclover/tsp/StreamSources.scala

```diff
@@ -327,3 +328,103 @@ case class InfluxDBSource(conf: InfluxDBInputConf, fieldsClasses: Seq[(Symbol, C
   override implicit def transformedTimeExtractor: TimeExtractor[Row] = RowIsoTimeExtractor(transformedTimeIndex, datetimeField)
 
 }
+
+object KafkaSource {
+
+  val log = Logger[KafkaSource]
+
+  def create(conf: KafkaInputConf)(implicit strEnv: StreamExecutionEnvironment): Either[ConfigErr, KafkaSource] =
+    for {
+      types <- KafkaService
+        .fetchFieldsTypesInfo(conf)
+        .toEither
+        .leftMap[ConfigErr](e => SourceUnavailable(Option(e.getMessage).getOrElse(e.toString)))
+      _ = log.info(s"Kafka types found: $types")
+      source <- StreamSource.findNullField(types.map(_._1), conf.datetimeField +: conf.partitionFields) match {
+        case Some(nullField) => KafkaSource(conf, types, nullField).asRight
+        case None            => InvalidRequest("Source should contain at least one non partition and datatime field.").asLeft
+      }
+    } yield source
+
+}
+
+case class KafkaSource(conf: KafkaInputConf, fieldsClasses: Seq[(Symbol, Class[_])], nullFieldId: Symbol)(
+  implicit @transient streamEnv: StreamExecutionEnvironment
+) extends StreamSource[Row, Symbol, Any] {
+
+  val log = Logger[KafkaSource]
+
+  def fieldsIdx = fieldsClasses.map(_._1).zipWithIndex
+  def fieldsIdxMap = fieldsIdx.toMap
+
+//  def fieldToEKey: Symbol => Int = { fieldId: Symbol =>
+//    fieldsIdxMap(fieldId)
+//  }
+
+  override def fieldToEKey: Symbol => Symbol = (x => x)
+
+  def timeIndex = fieldsIdxMap(conf.datetimeField)
+
+  def tsMultiplier = conf.timestampMultiplier.getOrElse {
+    log.info("timestampMultiplier in Kafka source conf is not provided, use default = 1000.0")
+    1000.0
+  }
+
+  implicit def extractor: ru.itclover.tsp.core.io.Extractor[org.apache.flink.types.Row, Symbol, Any] =
+    RowSymbolExtractor(fieldsIdxMap)
+  implicit def timeExtractor: ru.itclover.tsp.core.io.TimeExtractor[org.apache.flink.types.Row] =
+    RowTsTimeExtractor(timeIndex, tsMultiplier, conf.datetimeField)
+
+  val stageName = "Kafka input processing stage"
+
+  def createStream: DataStream[Row] = {
+    val consumer = KafkaService.consumer(conf, fieldsIdxMap)
+    consumer.setStartFromEarliest()
```

Comment:
I think we should pass starting offset as an argument here. Not critical, but need to be changed later, together with offset committing to improve reliability.

Reply:
Yes, it's temporary as for now, since it's much harder to test and check results without it.

Reply:
Ok, I see.
Let's add TODO here to not forget in future.